### ISSUE TITLE
better reporting of bad input arguments to jnl scripts (go files)

### ISSUE THREAD

**Issue by karlmsmith** [NOAA-PMEL/Ferret#1670]
**Reported by andrew wittenberg on 8 Apr 2016 22:55 UTC**
Create the following "test.jnl" that can receive a finite set of valid input arguments:

```
say $1%|a|b%
```

When we run this, we get:

```
 	NOAA/PMEL TMAP
 	FERRET v6.97 (beta/opt)
 	Linux 2.6.32-573.18.1.el6.x86_64 64-bit - 03/09/16
 	 8-Apr-16 18:45     

yes? can mode verify
yes? go test a
a
yes? go test b
b
yes? go test c
 **ERROR: invalid command: Valid value needed for argument $1
say $1%|a|b%
Command file, command group, or REPEAT execution aborted
           *** NOTE: Valid argument $1 choices are: a|b
yes? go test  
 **ERROR: invalid command: Valid value needed for argument $1
say $1%|a|b%
Command file, command group, or REPEAT execution aborted
           *** NOTE: Valid argument $1 choices are: a|b
```

Could we extend the "NOTE" to also mention what argument was received?  I.e. modify the last two examples to look something like this:

```
yes? go test c
 **ERROR: invalid command: Valid value needed for argument $1
say $1%|a|b%
Command file, command group, or REPEAT execution aborted
           *** NOTE: Argument $1 received was: c
           *** NOTE: Valid argument $1 choices are: a|b
yes? go test  
 **ERROR: invalid command: Valid value needed for argument $1
say $1%|a|b%
Command file, command group, or REPEAT execution aborted
           *** NOTE: Argument $1 was missing or empty.
           *** NOTE: Valid argument $1 choices are: a|b

```

This would greatly assist with debugging, e.g. when a long list of arguments is misaligned, or missing quotes, etc.

Migrated-From: http://dunkel.pmel.noaa.gov/trac/ferret/ticket/2398

**Comment by karlmsmith** [comment]
**Comment by ansley.b.manke on 8 Apr 2016 23:07 UTC**
This sounds like a nice idea. Implementing it shouldn't be difficult, though there will be details to handle with with argument substitutions and defaults.  I'll look into it.

**Comment by karlmsmith** [comment]
**Comment by ansley.b.manke on 9 Apr 2016 00:30 UTC**
Yes, this is easy.  With the one-line script shown above as testargs.jnl:


```
yes? go testargs
say $1%|a|b%
 **ERROR: invalid command: Valid value needed for argument $1
say $1%|a|b%
Command file, command group, or REPEAT execution aborted
           *** NOTE: Argument $1 was missing or empty.
           *** NOTE: Valid argument $1 choices are: a|b

yes? go testargs c
say $1%|a|b%
 **ERROR: invalid command: Valid value needed for argument $1
say $1%|a|b%
Command file, command group, or REPEAT execution aborted
           *** NOTE: Argument $1 received was: c
           *** NOTE: Valid argument $1 choices are: a|b
```

Now if the script has more arguments, and using substitution in argument 2:

```
say $1%|a|b% 

say $2%|a>1|c>3% 
```
 

Then send it various incorrect arguments:

```

yes? set mode ignore
yes? go testargs " " c
say $1%|a|b|*>c%
 **ERROR: invalid command: Valid value needed for argument $1
say $1%|a|b|*>c%
           *** NOTE: Argument $1 was missing or empty.
           *** NOTE: Valid argument $1 choices are: a|b|*>c
 
say $2%|a>1|c>3%
 !-> MESSAGE/CONTINUE 3
3


yes? go testargs a b
say $1%|a|b|*>c%
 !-> MESSAGE/CONTINUE a
a
 
say $2%|a>1|c>3%
 **ERROR: invalid command: Valid value needed for argument $2
say $2%|a>1|c>3%
           *** NOTE: Argument $2 received was: b
           *** NOTE: Valid argument $2 choices are: a>1|c>3

yes? go testargs b
say $1%|a|b|*>c%
 !-> MESSAGE/CONTINUE b
b
 
say $2%|a>1|c>3%
 **ERROR: invalid command: Valid value needed for argument $2
say $2%|a>1|c>3%
           *** NOTE: Argument $2 was missing or empty.
           *** NOTE: Valid argument $2 choices are: a>1|c>3

yes? ! With "set mode ignore" still in place,
yes? go testargs 
say $1%|a|b|*>c%
 **ERROR: invalid command: Valid value needed for argument $1
say $1%|a|b|*>c%
           *** NOTE: Argument $1 was missing or empty.
           *** NOTE: Valid argument $1 choices are: a|b|*>c
 
say $2%|a>1|c>3%
 **ERROR: invalid command: Valid value needed for argument $2
say $2%|a>1|c>3%
           *** NOTE: Argument $2 was missing or empty.
           *** NOTE: Valid argument $2 choices are: a>1|c>3
```

This will need a bit more testing, but there's no problem getting it in place.

**Comment by karlmsmith** [comment]
**Comment by Andrew.Wittenberg on 18 Oct 2016 16:22 UTC**
Thanks very much!  This looks to be implemented in v7.02.  Marking as fixed.
